com.avaloq.tools.ddk.xtext/src/com/avaloq/tools/ddk/xtext/modelinference/InferredModelAssociator.java

```diff
@@ -97,69 +106,76 @@ protected Adapter getOrInstall(final Resource resource) {
+    return getOrInstall(resource).getTargetToSourceMap();
   }
 
   /** {@inheritDoc} */
   @Override
   public void associate(final EObject sourceModelElement, final EObject inferredModelElement) {
     if (sourceModelElement != null) {
-      ListMultimap<EObject, EObject> map = getSourceToInferredModelMap(sourceModelElement.eResource());
-      map.put(sourceModelElement, inferredModelElement);
+      Resource resource = sourceModelElement.eResource();
+      Map<EObject, Set<EObject>> sourceToTargetMap = getSourceToTargetMap(resource);
+      putIntoSmallSetMap(sourceModelElement, inferredModelElement, sourceToTargetMap, false);
+      Map<EObject, Set<EObject>> targetToSourceMap = getTargetToSourceMap(resource);
+      putIntoSmallSetMap(inferredModelElement, sourceModelElement, targetToSourceMap, false);
     }
   }
 
   /** {@inheritDoc} */
   @Override
   public void associatePrimary(final EObject sourceModelElement, final EObject inferredModelElement) {
     if (sourceModelElement != null) {
-      ListMultimap<EObject, EObject> map = getSourceToInferredModelMap(sourceModelElement.eResource());
-      map.get(sourceModelElement).add(0, inferredModelElement);
+      Resource resource = sourceModelElement.eResource();
+      Map<EObject, Set<EObject>> sourceToTargetMap = getSourceToTargetMap(resource);
+      putIntoSmallSetMap(sourceModelElement, inferredModelElement, sourceToTargetMap, true);
+      Map<EObject, Set<EObject>> targetToSourceMap = getTargetToSourceMap(resource);
+      putIntoSmallSetMap(inferredModelElement, sourceModelElement, targetToSourceMap, true);
     }
   }
 
   /** {@inheritDoc} */
   @Override
   public Set<EObject> getInferredModelElements(final EObject sourceModelElement) {
-    if (sourceModelElement != null) {
-      final ListMultimap<EObject, EObject> sourceToInferredModelMap = getSourceToInferredModelMap(sourceModelElement.eResource());
-      final List<EObject> elements = sourceToInferredModelMap.get(sourceModelElement);
-      return newLinkedHashSet(elements);
-    } else {
+    if (sourceModelElement == null) {
       return Collections.emptySet();
     }
+    Map<EObject, Set<EObject>> map = getSourceToTargetMap(sourceModelElement.eResource());
+    Set<EObject> result = map.get(sourceModelElement);
+    return result != null ? result : Collections.emptySet();
   }
 
   /** {@inheritDoc} */
   @Override
   public Set<EObject> getSourceModelElements(final EObject inferredModelElement) {
-    if (inferredModelElement != null) {
-      ListMultimap<EObject, EObject> map = getSourceToInferredModelMap(inferredModelElement.eResource());
-      Set<EObject> sourceElements = newLinkedHashSet();
-      for (Map.Entry<EObject, EObject> entry : map.entries()) {
-        if (entry.getValue() == inferredModelElement) {
-          sourceElements.add(entry.getKey());
-        }
-      }
-      return sourceElements;
-    } else {
+    if (inferredModelElement == null) {

```

Comment:
coming back to targetToSource multimap... This method isn't used in asmd at all...

Reply:
Correct. I have not checked other downstream projects. Also here the API was basically copied from IJvmModelAssociations.